mvpa2/algorithms/group_clusterthr.py

```diff
@@ -0,0 +1,427 @@
+        # shortcuts
+        chunk_attr = self.params.chunk_attr
+        #
+        # Step 0: bootstrap maps by drawing one for each chunk and average them
+        # (do N iterations)
+        # this could take a lot of memory, hence instead of computing the maps
+        # we compute the source maps they can be computed from and then (re)build
+        # the matrix of bootstrapped maps either row-wise or column-wise (as
+        # needed) to save memory by a factor of (close to) `n_bootstrap`
+        # which samples belong to which chunk
+        chunk_samples = dict([(c, np.where(ds.sa[chunk_attr].value == c)[0])
+                                    for c in ds.sa[chunk_attr].unique])
+        # pre-built the bootstrap combinations
+        bcombos = [[random.sample(v, 1)[0] for v in chunk_samples.values()]
+                        for i in xrange(self.params.n_bootstrap)]
+        bcombos = np.array(bcombos, dtype=int)
+        # TODO implement parallel procedure as the estimation is independent
+        # across features
+        #
+        # Step 1: find the per-feature threshold that corresponds to some p
+        # in the NULL
+        thrsegs = []
+        segwidth = ds.nfeatures/self.params.n_blocks
+        # speed things up by operating on an array not a dataset
+        ds_samples = ds.samples
+        if __debug__:
+            debug('GCTHR',
+                  'Compute per-feature thresholds in %i blocks of %i features'
+                  % (self.params.n_blocks, segwidth))
+        thrmap = np.hstack( # merge across compute blocks
+                [get_thresholding_map(
+                    # one average map for every stored bcombo
+                    # this also slices the input data into feature subsets
+                    # for the compute blocks
+                    [np.mean(
+                        # get a view to a subset of the features
+                        # -- should be somewhat efficient as feature axis is
+                        # sliced
+                        ds_samples[sidx, segstart:segstart + segwidth],
+                        axis=0)
+                            for sidx in bcombos],
+                    self.params.feature_thresh_prob)
+                        # compute a partial threshold map for as mabny features
+                        # as fit into a compute block
+                        for segstart in xrange(0, ds.nfeatures, segwidth)])
+        # store for later thresholding of input data
+        self._thrmap = thrmap
+        #
+        # Step 2: threshold all NULL maps and build distribution of NULL cluster
+        #         sizes
+        #
+        cluster_sizes = Counter()
+        # recompute the bootstrap average maps to threshold them and determine
+        # cluster sizes
+        if 'mapper' in ds.a:
+            dsa = dict(mapper=ds.a.mapper)
+        else:
+            dsa = {}
+        if __debug__:
+            debug('GCTHR', 'Estimate NULL distribution of cluster sizes')
```

Comment:
nitpick.  IMHO imperative tone is a bit misleading here -- "Estimating ..." would be more representative and that is what we usually use in our debug msgs before something is going to happen